indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java

```diff
@@ -0,0 +1,194 @@
+    // check file gets uploaded to final HDFS path
+    Assert.assertTrue(fs.exists(expectedJarPath));
+    // check that the intermediate file gets deleted
+    Assert.assertFalse(fs.exists(new Path(intermediatePath, dummyJarFile.getName())));
+    // check file gets added to the classpath
+    Assert.assertEquals(expectedJarPath.toString(), job.getConfiguration().get(MRJobConfig.CLASSPATH_FILES));
+    Assert.assertEquals(dummyJarString, StringUtils.fromUtf8(IOUtils.toByteArray(fs.open(expectedJarPath))));
+  }
+
+  @Test
+  public void testIsSnapshot()
+  {
+    Assert.assertTrue(JobHelper.isSnapshot(new File("test-SNAPSHOT.jar")));
+    Assert.assertTrue(JobHelper.isSnapshot(new File("test-SNAPSHOT-selfcontained.jar")));
+    Assert.assertFalse(JobHelper.isSnapshot(new File("test.jar")));
+    Assert.assertFalse(JobHelper.isSnapshot(new File("test-selfcontained.jar")));
+    Assert.assertFalse(JobHelper.isSnapshot(new File("iAmNotSNAPSHOT.jar")));
+    Assert.assertFalse(JobHelper.isSnapshot(new File("iAmNotSNAPSHOT-selfcontained.jar")));
+
+  }
+
+  @Test
+  public void testConcurrentUpload() throws IOException, InterruptedException, ExecutionException, TimeoutException
+  {
+    final int concurrency = 10;
+    ExecutorService pool = Executors.newFixedThreadPool(concurrency);
+    // barrier ensures that all jobs try to add files to classpath at same time.
+    final CyclicBarrier barrier = new CyclicBarrier(concurrency);
+    final DistributedFileSystem fs = miniCluster.getFileSystem();
+    final Path expectedJarPath = new Path(finalClasspath, dummyJarFile.getName());
+    List<Future<Boolean>> futures = new ArrayList<>();
+
+    for (int i = 0; i < concurrency; i++) {
+      futures.add(
+          pool.submit(
+              new Callable()
+              {
+                @Override
+                public Boolean call() throws Exception
+                {
+                  int id = barrier.await();
+                  Job job = Job.getInstance(conf, "test-job-" + id);
+                  Path intermediatePathForJob = new Path(intermediatePath, "job-" + id);
+                  JobHelper.addJarToClassPath(dummyJarFile, finalClasspath, intermediatePathForJob, fs, job);
+                  // check file gets uploaded to final HDFS path
+                  Assert.assertTrue(fs.exists(expectedJarPath));
+                  // check that the intermediate file is not present
+                  Assert.assertFalse(fs.exists(new Path(intermediatePathForJob, dummyJarFile.getName())));
+                  // check file gets added to the classpath
+                  Assert.assertEquals(
+                      expectedJarPath.toString(),
+                      job.getConfiguration().get(MRJobConfig.CLASSPATH_FILES)
+                  );
+                  return true;
+                }
+              }
+          )
+      );
+    }
+
+    for (Future<Boolean> future : futures) {
```

Comment:
Then you don't really need to check the return value because it should bubble up any exceptions in the execution.